pkg/test/framework/components/echo/common/deployment/apps.go

```diff
@@ -0,0 +1,440 @@
+  location: MESH_EXTERNAL
+  resolution: DNS
+  endpoints:
+  - address: external.{{.Namespace}}.svc.cluster.local
+  ports:
+  - name: http-tls-origination
+    number: 8888
+    protocol: http
+    targetPort: 443
+  - name: http2-tls-origination
+    number: 8882
+    protocol: http2
+    targetPort: 443
+{{- range $i, $p := .Ports }}
+  - name: {{$p.Name}}
+    number: {{$p.ServicePort}}
+    protocol: "{{$p.Protocol}}"
+{{- end }}
+`, map[string]interface{}{"Namespace": externalNS.Name(), "Hostname": externalHostname, "Ports": serviceEntryPorts()})
+	if err != nil {
+		return err
+	}
+	if err := t.ConfigIstio().YAML(se).Apply(ns, resource.NoCleanup); err != nil {
+		return err
+	}
+	return nil
+}
+
+type External struct {
+	// Namespace where external echo app will be deployed
+	Namespace namespace.Instance
+
+	// Echos app to be used by tests, with no sidecar injected
+	Echos echo.Instances
+}
+
+func (e *External) loadValues(echos echo.Instances) {
+	e.Echos = match.ServiceName(model.NamespacedName{Name: ExternalSvc, Namespace: e.Namespace.Name()}).GetMatches(echos)
+}
+
+type EchoDeployments struct {
+	// Ingressgateway instance
+	Ingress   ingress.Instance
+	Ingresses ingress.Instances
+
+	// Echo deployments in the first namespace
+	Ns1 Namespace
+
+	// Echo deployments in the second namespace.
+	Ns2 Namespace
+
+	// Ns1AndNs2 combines all deployments in both Ns1 and Ns2.
+	Ns1AndNs2 echo.Services
+
+	// External (out-of-mesh) deployments
+	External External
+
+	// All echo instances.
+	All echo.Services
+}
+
+func (d *EchoDeployments) loadValues(t resource.Context, echos echo.Instances) error {
+	d.All = echos.Services()
+
+	if err := d.Ns1.loadValues(t, echos, d.Ns2.Namespace, d.External.Namespace); err != nil {
+		return err
+	}
+	if err := d.Ns2.loadValues(t, echos, d.Ns1.Namespace, d.External.Namespace); err != nil {
+		return err
+	}
+
+	d.Ns1AndNs2 = d.Ns1.All.Instances().Append(d.Ns2.All.Instances()).Services()
+
+	d.External.loadValues(echos)
+	return nil
+}
+
+const (
+	PodASvc          = "a"
+	PodBSvc          = "b"
+	PodCSvc          = "c"
+	PodTproxySvc     = "tproxy"
+	VMSvc            = "vm"
+	HeadlessSvc      = "headless"
+	StatefulSetSvc   = "statefulset"
+	ProxylessGRPCSvc = "proxyless-grpc"
+	NakedSvc         = "naked"
+	ExternalSvc      = "external"
+	DeltaSvc         = "delta"
+
+	externalHostname = "fake.external.com"
+)
+
+func serviceEntryPorts() []echo.Port {
+	var res []echo.Port
+	for _, p := range ports.All().GetServicePorts() {
+		if strings.HasPrefix(p.Name, "auto") {
+			// The protocol needs to be set in common.EchoPorts to configure the echo deployment
+			// But for service entry, we want to ensure we set it to "" which will use sniffing
+			p.Protocol = ""
+		}
+		res = append(res, p)
+	}
+	return res
+}
+
+func SetupApps(t resource.Context, i istio.Instance, apps *EchoDeployments) error {
+	var err error
+	apps.Ns1.Namespace, err = namespace.New(t, namespace.Config{
+		Prefix: "echo1",
```

Comment:
really nit but can we keep this as `echo`? I have `echo` in my muscle memory...

Reply:
We need the prefix for each namespace to be unique for test names. I guess it could be `echo` and `echo1`, but the variables are Ns1 and Ns2 ... a bit odd.